File: src/main/java/com/matag/game/adminclient/AdminClient.java
package com.matag.game.adminclient;

import static java.util.Collections.singletonList;

import com.matag.adminentities.DeckInfo;
import com.matag.adminentities.FinishGameRequest;
import com.matag.adminentities.PlayerInfo;
import com.matag.game.config.ConfigService;
import com.matag.game.player.Player;
import com.matag.game.security.SecurityToken;
import lombok.AllArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Profile;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

@Profile("!test")
@AllArgsConstructor
@Component
public class AdminClient {
  private static final Logger LOGGER = LoggerFactory.getLogger(AdminClient.class);

  private final ConfigService configService;

  public DeckInfo getDeckInfo(SecurityToken token) {
    return get(token, "/game/active-deck", DeckInfo.class);
  }

  public PlayerInfo getPlayerInfo(SecurityToken token) {
    return get(token, "/player/info", PlayerInfo.class);
  }

  public void finishGame(String gameId, Player winner) {
    try {
      adminPost("/game/" + gameId + "/finish", new FinishGameRequest(winner.getToken().getAdminToken()), Object.class);
    } catch (Exception e) {
      LOGGER.error("Could not finish the game.", e);
    }
  }

  private <T> T get(SecurityToken token, String url, Class<T> responseType) {
    return exchange(token, url, HttpMethod.GET, null, responseType).getBody();
  }

  private <T> T adminGet(String url, Class<T> responseType) {
    return exchange(null, url, HttpMethod.GET, null, responseType).getBody();
  }

  private <T> T post(SecurityToken token, String url, Object body, Class<T> responseType) {
    return exchange(token, url, HttpMethod.POST, body, responseType).getBody();
  }

  private <T> T adminPost(String url, Object body, Class<T> responseType) {
    return exchange(null, url, HttpMethod.POST, body, responseType).getBody();
  }

  private <T> ResponseEntity<T> exchange(SecurityToken token, String url, HttpMethod method, Object request, Class<T> responseType) {
    var restTemplate = new RestTemplate();
    var headers = new HttpHeaders();
    headers.setAccept(singletonList(MediaType.APPLICATION_JSON));

    if (token != null) {
      headers.set("session", token.getAdminToken());
    } else {
      headers.set("admin", configService.getAdminPassword());
    }

    var entity = new HttpEntity<>(request, headers);

    return restTemplate.exchange(configService.getMatagAdminUrl() + url, method, entity, responseType);
  }
}

File: src/main/kotlin/com/matag/game/adminclient/AdminClient.kt
package com.matag.game.adminclient

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.matag.adminentities.DeckInfo
import com.matag.adminentities.FinishGameRequest
import com.matag.adminentities.PlayerInfo
import com.matag.game.config.ConfigService
import com.matag.game.player.Player
import com.matag.game.security.SecurityToken
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Profile
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate

@Profile("!test")
@Component
open class AdminClient(
    private val configService: ConfigService
) {
    open fun getDeckInfo(token: SecurityToken?): DeckInfo {
        return get(token, "/game/active-deck", DeckInfo::class.java)
    }

    open fun getPlayerInfo(token: SecurityToken?): PlayerInfo {
        return get(token, "/player/info", PlayerInfo::class.java)
    }

    fun finishGame(gameId: String?, winner: Player) {
        try {
            adminPost(
                "/game/$gameId/finish",
                FinishGameRequest(winner.token.adminToken),
                Any::class.java
            )
        } catch (e: Exception) {
            LOGGER.error("Could not finish the game.", e)
        }
    }

    private fun <T> get(token: SecurityToken?, url: String, responseType: Class<T>): T {
        return exchange(token, url, HttpMethod.GET, null, responseType)
    }

    private fun <T> adminGet(url: String?, responseType: Class<T>): T {
        return exchange(null, url, HttpMethod.GET, null, responseType)
    }

    private fun <T> post(token: SecurityToken?, url: String, body: Any, responseType: Class<T>): T {
        return exchange(token, url, HttpMethod.POST, body, responseType)
    }

    private fun <T> adminPost(url: String?, body: Any?, responseType: Class<T>): T {
        return exchange(null, url, HttpMethod.POST, body, responseType)
    }

    private fun <T> exchange(
        token: SecurityToken?,
        url: String?,
        method: HttpMethod,
        request: Any?,
        responseType: Class<T>
    ): T {
        val restTemplate = RestTemplate()
        val headers = HttpHeaders()
        headers.accept = listOf(MediaType.APPLICATION_JSON)

        if (token != null) {
            headers.set("session", token.adminToken)
        } else {
            headers.set("admin", configService.adminPassword)
        }

        val entity = HttpEntity(request, headers)

        val response = restTemplate.exchange(
            configService.matagAdminInternalUrl + url,
            method,
            entity,
            String::class.java
        )
        return OBJECT_MAPPER.readValue<T>(response.getBody(), responseType)
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(AdminClient::class.java)
        private val OBJECT_MAPPER: ObjectMapper =
            jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
    }
}
